File: Lab3/Lab3/src/KI304/Moh/Lab3/Sea.java
package KI304.Moh.Lab3;

/**
 * The {@code Sea} class represents a sea, extending the {@code Reservoir} class
 * and implementing the {@code Fishing} interface. It includes additional properties
 * specific to a sea, such as the number of sharks and fishing allowances.
 *
 * @author Marian Moh
 * @version 1.0
 * @since 2023-11-25
 */
public class Sea extends Reservoir implements Fishing {

    /** The number of sharks in the sea. */
    public int numberOfSharks;

    /** Flag indicating whether shark fishing is allowed in the sea. */
    public boolean sharkFishingIsAllowed;

    /**
     * Default constructor for the {@code Sea} class.
     */
    public Sea() {
    }

    /**
     * Parameterized constructor for the {@code Sea} class.
     * Initializes the sea with specified properties and checks fishing allowances.
     *
     * @param name            The name of the sea.
     * @param amountOfWater   The amount of water in the sea.
     * @param waterTemperature The temperature of the water in the sea.
     * @param numberOfSharks  The initial number of sharks in the sea.
     */
    public Sea(String name, int amountOfWater, int waterTemperature, int numberOfSharks) {
        super(name, amountOfWater, waterTemperature);
        this.numberOfSharks = numberOfSharks;
        isFishingAllowed();
    }

    /**
     * Displays information about the sea, including its name, amount of water,
     * water temperature, number of sharks, current season, and fishing allowances.
     */
    @Override
    public void showInfo() {
        log("\nName: " + super.getName());
        log("Amount of water: " + super.getAmountOfWater());
        log("Water temperature: " + super.getWaterTemperature());
        log("Number of sharks: " + numberOfSharks);

        if(!seasonIsUnknown()) {
            log("Season is: " + super.getSeason());
        }

        if(!sharkFishingIsAllowed) {
            log("- Shark fishing is not allowed!!!");
        } else {
            log("Shark fishing is allowed");
        }
    }

    /**
     * Sets the properties of the sea for the winter.
     * Logs the change and displays information about the sea.
     */
    @Override
    public void winterIsComing() {
        setSeason(Season.WINTER);
        setWaterTemperature(5);
        showInfo();
    }

    /**
     * Sets the properties of the sea for the spring.
     * Logs the change, adjusts the amount of water, and displays information about the sea.
     */
    @Override
    public void springIsComing() {
        setSeason(Season.SPRING);
        setWaterTemperature(14);
        setAmountOfWater((int) (getAmountOfWater() * 1.2));
        showInfo();
    }

    /**
     * Sets the properties of the sea for the summer.
     * Logs the change, adjusts the amount of water, and displays information about the sea.
     * If the amount of water is less than 500000, the sea dries up.
     */
    @Override
    public void summerIsComing() {
        setSeason(Season.SUMMER);
        setWaterTemperature(19);
        setAmountOfWater((int) (getAmountOfWater() * 0.8));
        if(getAmountOfWater() < 500000) {
            dry();
        }
        showInfo();
    }

    /**
     * Sets the properties of the sea for the autumn.
     * Logs the change and displays information about the sea.
     */
    @Override
    public void autumnIsComing() {
        setSeason(Season.AUTUMN);
        setWaterTemperature(16);
        showInfo();
    }

    /**
     * Performs the action of fishing in the sea.
     * If shark fishing is allowed, reduces the number of sharks,
     * checks if further fishing is allowed, and logs the result.
     */
    @Override
    public void fish() {
        if(sharkFishingIsAllowed) {
            numberOfSharks = (int) (numberOfSharks * 0.75);
            isFishingAllowed();
            log("\nAfter fishing the number of sharks is: " + numberOfSharks);
        } else {
            log("\n- Shark fishing is not allowed! Too few sharks!!!");
        }
    }

    /**
     * Performs the action of feeding in the sea.
     * Increases the number of sharks, checks if fishing is allowed,
     * and logs the result.
     */
    @Override
    public void feed() {
        numberOfSharks = (int) (numberOfSharks * 1.1);
        isFishingAllowed();
        log("\nAfter feeding the number of sharks is: " + numberOfSharks);
    }

    /**
     * Checks whether fishing is currently allowed in the sea.
     * Updates the fishing allowance flag based on the number of sharks.
     */
    @Override
    public void isFishingAllowed() {
        sharkFishingIsAllowed = numberOfSharks > 2000;
    }
}

File: Lab9/Lab9/sea.py

from reservoir import Reservoir

class Sea(Reservoir):
    # Initialize the Sea object, inheriting from Reservoir.
    def __init__(self, name, amountOfWater, waterTemperature, numberOfSharks):
        super().__init__(name, amountOfWater, waterTemperature)
        self.__numberOfSharks = numberOfSharks
        self.__sharkFishingIsAllowed = numberOfSharks > 2000

    # Display information about the sea, including inherited reservoir information and
    # specific sea attributes such as the number of sharks and shark fishing allowance.
    def showInfo(self):
        super().showInfo()
        print("Number of sharks: ", self.__numberOfSharks)

        if self.__sharkFishingIsAllowed:
            print("Shark fishing is allowed!")
        else:
            print("- Shark fishing is not allowed!!!")

    # Perform fishing in the sea, reducing the number of sharks and updating fishing allowance.
    def fish(self):
        if self.__sharkFishingIsAllowed:
            self.__numberOfSharks = int(self.__numberOfSharks * 0.75)
            self.isFishingAllowed()
            print(f"\nAfter fishing there are {self.__numberOfSharks} sharks")
        else:
            print(f"\n- Shark fishing is not allowed! Only {self.__numberOfSharks} sharks!!!")

    # Feed the sharks in the sea, increasing the number of sharks and updating fishing allowance.
    def feed(self):
        self.__numberOfSharks = int(self.__numberOfSharks * 1.1)
        self.isFishingAllowed()
        print(f"\nAfter feeding there are {self.__numberOfSharks} sharks")

    # Check whether shark fishing is allowed based on the updated number of sharks.
    def isFishingAllowed(self):
        self.__sharkFishingIsAllowed = self.__numberOfSharks > 2000
